MovingBlocks/Terasology
# feat: GooeyCLI - replacement utility scripting system based on PicoCLI.

Still a proof of concept, with a limited amount of old functionality converted to showcase the new approach. Relies on a `groovyw` tweak (https://github.com/MovingBlocks/groovy-wrapper/pull/5), renames the scripts (`gw` is < 1/3rd of `groovyw` ! Think of the savings!), and lets [PicoCLI](https://picocli.info/) worry about all the CLI structuring. Also experimenting with https://github.com/Terasology/Index/pull/7

### Contains

Overhaul of our utility scripting. PicoCLI takes away _all_ the awkwardness surrounding argument parsing and such, plus automatically generates usage text, even at every level, a la `kubectl`, `docker`, `git`, etc. 

Key part is defining commands and options inside small classes using PicoCLI. A command class can define a link to sub-command classes, which in turn can have their own and so on (`gw` -> `module` -> `get `..). From anywhere you can mix-in other classes containing simple definitions for options, so you only need to define a given option (like `--origin`) in one place then just mix it in wherever needed. If something isn't relevant for a given combination just don't link it. Easy!

* Replaces a ton of messy code - `util.groovy` is no more, the weird loading of files as text to then turn them into Groovy scripts dynamically is gone, and you can write way cleaner, well, everything.
* Adding new commands is trivial rather than painful, and it is easy to add them at any level with a single small file. Options / Parameters can be added separately as traits / mix-ins, then plugged in wherever
* Is technically in Groovy, but remains largely easy Java without a lot of elaborate Groovy'isms. I did try for a bit to get plain Groovy scripts to work (cuts down even more boilerplate code) but it didn't seem worth the hassle / weirdness / quirks in trying to use the setup with our Groovy Wrapper
* A `help` / `-h` / `--help` command is available at every level, detailing everything automatically. Extra examples can easily be added beyond the auto-generated stuff
* Working text colors on Windows - previous ANSI coloring only works on Linux/Mac, and leaves control characters in the output on Windows. Some stuff like underlining still doesn't work on Windows, but colors do 🌈 
* Docs for PicoCLI are pretty great, it seems very mature, and also has lots of nice examples. Plus it is apparently embedded within Groovy itself!

### Diagram!

![PicoCLI](https://user-images.githubusercontent.com/1063833/86503700-1687a380-bd76-11ea-85e5-47c814561f73.png)

The green stuff can actually be run, everything else is supporting (some sub-commands require additional sub-commands - you can run `gw module` but will get an error about that being meaningless on its own). Diagram doesn't cover all use cases and not everything listed is done yet.

### How to test

* Check out PR
* Run `gw` commands in the root of your workspace

That's it, really! Everything else should just work, no extra steps needed. Not a lot is actually functional yet

* Try to use the `help` command in various places. Or nothing at all and see what happens
* `gw module get ModuleName` should work normally, you can add `-o` / `--origin` to vary where to get modules from. The old `gradle.properties` global override for different GitHub locations should also work, if set (example: `alternativeGithubHome=Nanoware`)
  * You should be able to get multiple modules in one go, but `recurse` and other stuff isn't in yet
* `gw init` says some things, including colored output!
* Try putting the `-o` flag in different places, using different commands
* `gw module refresh` should do the `build.gradle` update to all modules, demonstrates a sub-command defined specifically within an item type - sort of like a private command to that parent, rather than write it in its own file, when it won't make sense anywhere else

### Outstanding before merging

- [ ] Sanity checks for architecture and such (I feel like I code so rarely I'm forgetting all the patterns and probably could have done this better), plus actually confirm if we want to do this - #4035
- [ ] Finish converting old functionality
- [ ] See if the PicoCLI and jansi jars in the Gradle Wrapper can just be added to the classpath to avoid their `@Grab`s
- [ ] May well move to a future goal but: extract the utility scripting to its own repo and solely `@Grab` a built jar from there. In most cases users don't really need the scripts themselves, and I'm sure we could come up with a way to sub in source when you want to edit it, like with other items. **Benefit:** Use the exact same setup for both Terasology and DestSol (and more?), limit the amount of code in the engine repo, possibly version things better?

### Future goals

While not a goal with this PR I'm hoping to build more advanced functionality on top of it, partly via my day job where I'm trying to do something similar in a different use case (enterprise CI/CD / app dev)

* Add back in command combinations that ask the user for input? Unsure how useful that really is, so rather wait till later
* Look into hooking the command auto-completion in (even if it might not work on Windows)
* A promote / release command that'll tag a repo on GitHub. Possibly deal with a GitHub release? Or would that be Jenkins? #4035 again
* A `fork` command that'll create a fork on GitHub of a target item
* A command that'll create a normal repo remotely on GitHub (not a fork). Name?
* A `pr` command that'll create a PR automagically
* `bulk` commands to handle large-scale modifications, such as an engine change needing minor tweaks in a dozen or two modules (prep branches, bulk commit, push to GitHub, make / merge PRs if desired, possibly add markup somewhere to tell Jenkins to test them all together)
* Possibly minor reporting options? Related to open PRs, branches, and such. Jenkins X does some neat things
* Possibly utility scripting to help manage teams on GitHub, adding new users easily, standardize user lists, make labels the same in different repos,etc
* Git adapters for other providers (GitLab, etc) - not necessarily very useful for us here but I need it for work anyway, so here eventually we might be able to support modules on other Git providers seamlessly?
* Some ways later, maybe: use some of this logic remotely via Jenkins and/or ChatOps (use the commands on chat to create repos for new users, release things, detail forks, etc)
* Possibly use PicoCLI itself _inside_ the game - could it power our in-game console? We already have code overlap there between TS and DS, maybe we could have a game console library both projects could use. Been wondering about this since I started digging around, and recently actually found https://github.com/Rubydesic/MinecraftPicocli - not sure if that's in-game console or server console 🤔 
* Maybe support `wiki` items or even `gh-pages` items if we end up using those
